app/controllers/groups/members_controller.rb

```diff
@@ -31,6 +31,28 @@ def join
     end
   end
 
+  def add
+    @group = Group.find(params[:id])
+    authorize @group, :add?
+    @user = User.find_by_username(params[:username])
+    if @user.nil?
+      redirect_to(members_group_path(@group), :alert => "No user found with username \"#{params[:username]}\"")
+    elsif @group.members.exists?(@user)
+      redirect_to(members_group_path(@group), :alert => "#{@user.username} is already a member of this group")
+    else
+      @group.join(@user)
+      redirect_to(members_group_path(@group), :notice => "#{@user.username} has been added to this group")
+    end
+  end
+
+  def remove
+    @group = Group.find(params[:id])
+    authorize @group, :remove?
+    @user = User.find(params[:user_id])
+    @group.members.delete(@user)
+    redirect_to(members_group_path(@group), :notice => "#{@user.username} has been removed from this group")
+  end
+
```

Comment:
I think the new add/remove methods should be before the existing join method rather than between join/leave (unless there is a particular reason to put them there?)

Reply:
Sure, I'm fine with that.